org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java

```diff
@@ -162,19 +196,23 @@ private int generate(CMElementDeclaration elementDeclaration, String prefix, boo
 		snippetIndex = generate(attributes, level, snippetIndex, xml, elementDeclaration.getLocalName());
 		// Elements children
 		Collection<CMElementDeclaration> children = elementDeclaration.getElements();
+		boolean childGenerated = false;

```

Comment:
I suggest that you provide a new method `getRequiredElements()` to write:

```java
Collection<CMElementDeclaration> children = generateOnlyRequired  ? elementDeclaration.getRequiredElements(); : elementDeclaration.getElements();
```

For XSD and DTD this method could be implemented by looping getElements()  and check if isOptional return false (like today).

For RelaxNG, you compute the requiredElementsName (like it is done today) and you loop for each name and build the list with CMRelaxNGElementDeclaration.